docs/content/preview/architecture/transactions/read-restart-error.md

```diff
@@ -32,6 +32,12 @@ A detailed scenario that explains how clock skew can result in the above mention
+  * It should not return the data if the write is performed in the future (that is, after the read point had been chosen) **and** had a write timestamp later than the read point. Because if that is allowed, everything written in future would be trivially visible to the read. Note that the latter condition is important because it is okay to return data that was written after the read point was picked, if the write timestamp was earlier than the read point (this doesn't break any consistency or isolation guarantees).
 
 * If node `N2` finds writes in the range `(T1, T1+max_clock_skew]`, to avoid breaking the strong guarantee that a reader should always be able to read what was committed earlier, and to avoid reading data with a later write timestamp which was also actually written after the read had been issued, node `N2` raises a `Read restart` error.
+
+## Troubleshooting
+
+1. Implement retry logic in the application. Application retries can help with mitigating read restart errors. Moreover, there are other ways in which a statement or a transaction may fail such as transaction conflicts or infrastructure failures. Therefore, a retry mechanism is strongly recommended for a cloud-native, distributed database such as YugabyteDB.
+2. Use SERIALIZABLE, READ ONLY, DEFERRABLE mode whenever possible. Read restart errors usually occur when the query is a SELECT statement with a large output footprint. These are better run with the above-mentioned transaction characteristics anyway. Using DEFERRABLE will avoid a read restart error altogether. However, the tradeoff is that the statement waits out the clock skew before reading the data. This is not an issue for large SELECT statements. Tip: use `SET SESSION CHARACTERISTICS AS TRANSACTION ISOLATION LEVEL SERIALIZABLE READ ONLY DEFERRABLE;` to change the default transaction properties.
+3. Using read only, deferrable transactions is not always feasible, either because the query is not read only, or because it is not acceptable to wait out the clock skew. In these cases, try increasing the value of `ysql_output_buffer_size`. This will enable YugabyteDB to retry the query internally on behalf of the user. As a tradeoff, more memory is used by the YSQL backend processes, increasing the risk of out-of-memory errors. Be aware that this is not a silver bullet. For example, the COPY command can still raise a read restart error even though the command has a one line output. Increasing `ysql_output_buffer_size` is not useful in this scenario. The application must retry the COPY command instead. Another example is DMLs such as INSERT/UPDATE/DELETE. These do not have enough output to overflow the buffer size. However, when these statements are executed as part of an explicit transaction block (e.g. BEGIN; ... INSERT ... COMMIT;), a read restart error cannot be retried internally by YugabyteDB. The onus is on the application to ROLLBACK and retry the transaction.
```

Comment:
```suggestion
- Using read only, deferrable transactions is not always feasible, either because the query is not read only, or because it is not acceptable to wait out the clock skew. In these cases, try increasing the value of the `ysql_output_buffer_size` flag. This enables YugabyteDB to retry the query internally on behalf of the user. As a tradeoff, the YSQL backend processes use more memory, increasing the risk of out-of-memory errors.

  Be aware that this is not a silver bullet. For example, the COPY command can still raise a read restart error even though the command has a one line output. Increasing `ysql_output_buffer_size` is not useful in this scenario. The application must retry the COPY command instead. Another example is DMLs such as INSERT/UPDATE/DELETE. These do not have enough output to overflow the buffer size. However, when these statements are executed as part of an explicit transaction block (for example, BEGIN; ... INSERT ... COMMIT;), a read restart error cannot be retried internally by YugabyteDB. The onus is on the application to ROLLBACK and retry the transaction.
```

Reply:
maybe the flag should also be added to the yb-tserver reference, doesn't appear to be documented

Reply:
Using read only, deferrable transactions is not always feasible, either because the query is not read only, or the query is part of a read-write transaction or because ...

This will enable YugabyteDB to retry the query internally on behalf of the user. As long as the output of a statement hasn't crossed `ysql_output_buffer_size ` to result in flushing partial data to the external client, the YSQL query layer retries read restart errors for all statements in a Read Committed transaction block, for the first statement in a Repeatable Read transaction block and for any standalone statement outside a transaction block.

"However, when these statements are executed as part of an explicit transaction block (e.g. BEGIN; ... INSERT ... COMMIT;), a read restart error cannot be retried internally by YugabyteDB" -- this statement isn't right, we do retry these statements in RC txns and if they are the first statement in a RR txn.

Reply:
Filed https://github.com/yugabyte/yugabyte-db/issues/23232 for documentation of ysql_output_buffer_size flag.

Reply:
What about the comment by Piyush "- this statement isn't right, we do retry these statements in RC txns and if they are the first statement in a RR txn"

Reply:
@ddhodge  I've changed the wording to say REPEATABLE READ transaction and that the statement is in the middle of the transaction.